codec/gcm.go

```diff
@@ -29,23 +61,42 @@ func NewGCMSession(key [32]byte) (*GCMSession, error) {
 	if err != nil {
 		return &GCMSession{}, fmt.Errorf("creating gcm cipher: %v", err)
 	}
-	return &GCMSession{
-		gcm:       gcm,
-		wRand:     rand.New(rand.NewSource(int64(binary.BigEndian.Uint64(key[:8])))),
-		wBuf:      make([]byte, gcm.NonceSize()),
-		nonceSize: gcm.NonceSize(),
-	}, nil
+
+	gcmSession := &GCMSession{
+		gcm:        gcm,
+		readNonce:  &gcmNonce{},
+		writeNonce: &gcmNonce{},
+	}
+
+	if bytes.Compare(self[:], remote[:]) < 0 {
+		gcmSession.writeNonce.top = math.MaxUint32
+		gcmSession.writeNonce.bottom = math.MaxUint64
+		gcmSession.writeNonce.countDown = true
+	} else {
+		gcmSession.readNonce.top = math.MaxUint32
+		gcmSession.readNonce.bottom = math.MaxUint64
+		gcmSession.readNonce.countDown = true
+
+	}
+	return gcmSession, nil
 }
 
 func GCMEncoder(session *GCMSession, enc Encoder) Encoder {
 	return func(w io.Writer, buf []byte) (int, error) {
-		_, err := session.wRand.Read(session.wBuf)
+		nonceBufTop, err := surge.ToBinary(session.writeNonce.top)
 		if err != nil {
-			return 0, fmt.Errorf("generating randomness: %v", err)
+			return 0, fmt.Errorf("marshaling nonce to bytes: %v", err)
 		}
-		n, err := enc(w, append(session.wBuf, session.gcm.Seal(nil, session.wBuf, buf, nil)...))
+		nonceBufBottom, err := surge.ToBinary(session.writeNonce.bottom)
 		if err != nil {
-			return n, fmt.Errorf("encoding sealed data: %v", err)
+			return 0, fmt.Errorf("marshaling nonce to bytes: %v", err)
+		}
+		nonceBuf := append(nonceBufTop, nonceBufBottom...)
```

Comment:
This is quite an inefficient way to do this. Lots of allocation and copying is happening (`ToBinary` is expensive). Instead, consider using something like:

```
nonceBuf := [12]byte{}
binary.BigEndian.PutUint32(nonceBuf[:], session.writeNonce.top)
binary.BigEndian.PutUint64(nonceBuf[4:], session.writeNonce.bottom)
```

Reply:
Same applies when decoding/opening.